Various Contract Enhancements

apbendi opened the pull request:
- [x] Change how token amounts are stored #98 
- [x] Add chainid & version to meta withdrawal signature #74 
- [x] Natspec comments #18

closes #98 
closes #74 
closes #18

apbendi commented:
@mds1 sometime tomorrow afternoon I plan on taking a look at how my changes have impacted `umbra-js` and fixing that up, but iff you have the bandwidth and wanted an umbra task before then, feel free to take a look. Would also welcome feedback on the changes so far.

mds1 commented on contracts/contracts/Umbra.sol:
```diff
@@ -69,72 +133,99 @@ contract Umbra is Ownable {
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   * @param _sponsor Address which is compensated for submitting the withdrawal tx
+   * @param _sponsorFee Amount of the token to pay to the transfer
+   * @param _v ECSDA signature prefix
+   * @param _r ECSDA signature x-coordinate
+   * @param _s ECSDA signature y-coordinate
```
These definitions aren't quite correct (seems you were thinking of uncompressed public keys), we can either use the ethers phrasing from [here](https://docs.ethers.io/v5/single-page/#/v5/api/utils/bytes/-%23-Signature), so:

```
   * @param _v ECSDA signature component: Parity of the `y` coordinate of point `R`
   * @param _r ECSDA signature component: x-coordinate of `R`
   * @param _s ECSDA signature component: `s` value of the signature
```

Or alternatively just lump everything into a `@dev` comment


```
   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
   * @dev _v, _r, _s are from a valid secp256k1 signature from owner
   * @param _stealthAddr The stealth address whose token balance will be withdrawn
   * @param _acceptor Address where withdrawn funds should be sent
   * @param _tokenAddr Address of the ERC20 token being withdrawn
   * @param _sponsor Address which is compensated for submitting the withdrawal tx
   * @param _sponsorFee Amount of the token to pay to the transfer
   * @param _v ECSDA signature v-component
   * @param _r ECSDA signature r-component
   * @param _s ECSDA signature s-component

apbendi replied:
Nice, thanks for the heads up here. I like the ethers phrasing 👍

mds1 reviewed (changes requested):
Looks good! I added a lot of comments but most are small!

Since this PR is generically called "Various Contract Enhancements", I've added some other ideas below. Feel free to move them to a future issue if that makes more sense

1. Issue #18 also mentions checking out solidity docgen to autogenerate docs from the natspec comments. Since we aren't doing this, we might want to create a separate issue to track documentation in general 

2. At this point 0.6.12 is a bit old. We could bump to 0.7.6 or even 0.8.2 since the latest version of `@openzeppelin/contracts` supports it (this means we can remove SafeMath since those checks are built in to 0.8.x, which will reduce gas cost)

3. On the topic of saving gas, we could also get rid of `Ownable.sol` and implement it ourselves. `Owable` pulls in `Context.sol` to use `_msgSender()` which we don't need and makes the `onlyOwner` modifier + ownership transfer slightly more expensive. Ultimately `_msgSender()` just returns `msg.sender`, but still adds a little overhead. This is probably not too important of a change so I don't feel strongly here

4. We don't need `import "@openzeppelin/contracts/token/ERC20/IERC20.sol";` since the `SafeERC20` import already imports `IERC20`

5. In `address constant ETH_TOKEN_PLACHOLDER = address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE);`, we can remove the `address()` cast

6. I can't run tests, presumably since running `yarn` at the root fails since `umbra-js` is broken, though the actual error I get with `yarn test` is `Error HH108: Cannot connect to the network localhost. Please make sure your node is running, and check your internet connection and networks config`, which seems unrelated.

mds1 commented:
@apbendi Just pushed commit so umbra-js and the frontend account for the new contract changes. We'll need to deploy a new Rinkeby contract for the frontend, and we'll also need to update ENS support in the frontend, so we can batch all the frontend changes into one PR later if we want

------------------------

@wildmolasses Any idea why the `/contracts` folder doesn't seem to automatically run Prettier on staged files? If you switch to this branch's latest commit and run `yarn prettier`, you'll see a lot of the files Ben changes are formatted, i.e they weren't formatted on commit. I noticed the `umbra-js/package.json` has `"precommit": "lint-staged"`, but `contracts/package.json` doesn't, so I'm guessing that's why.

Similar issue has come up before in this repo (https://github.com/ScopeLift/umbra-protocol/pull/95#issuecomment-782731019) so would love to get this Prettier issue fixed for good!

wildmolasses commented:
> I noticed the `umbra-js/package.json` has `"precommit": "lint-staged"`, but `contracts/package.json` doesn't, so I'm guessing that's why.

@mds1 Yes, exactly! Does PR owner mind adding `"precommit": "lint-staged",` to contracts/package.json scripts? Conveniently, the contracts folder already has a .lintstagedrc.js file, so this should be the only change needed.

apbendi commented on contracts/contracts/Umbra.sol:
```diff
@@ -10,28 +10,55 @@ import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
-    address indexed receiver,
-    uint256 indexed amount,
-    address indexed token,
+    address indexed receiver, // stealth address
```
I'm not sure I follow how the uniqueness of the stealth address negates the need to index it. Indexing makes it possible to filter logs on the parameter. So the question is, "will I want to filter event logs on a given stealth address." I think there are cases where we might want to, for example #34, where we might want to filter on a pre-computed stealth address associated with a repeat stealth payment.

apbendi commented on contracts/contracts/Umbra.sol:
```diff
@@ -10,28 +10,55 @@ import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+  /// @notice Emitted when a token is withdrawn
+  event TokenWithdrawal(
+    address indexed receiver, // stealth address
+    address indexed acceptor, // destination of funds
```
Similar to above, I wonder if there might be some need in the future to filter on the destination where funds were sent. I can't think of an example off the top of my head, but I can imagine the need might exist. The gas overhead for indexing is 375 so the question is whether it's worth it to be able to have the option to do this filtering or not.

mds1 replied:
Now it's 8.2 cents! 😛 

So between these two parameters we're looking at 16 cents per transaction, which sounds small, but (1) it will likely only go up over time, and (2) making everyone pay more for an unknown future use case doesn't seem ideal given the state of gas prices.

Personally I learn towards saving the gas, and if someone needs to filter by stealth address they can just use a third party service to help with parsing the logs, such as Dune

apbendi replied:
Here's a usecase I thought of: if you're writing a contract to act as a withdrawal hook, you might want to filter on all 
withdrawals where your contract was the acceptor.

That said, I completely agree that gas prices are likely to continue going up and we should be super sensitive to increasing gas costs needlessly.

> if someone needs to filter by stealth address or acceptor they can just use a third party service to help with parsing the logs, such as Dune

This would be an argument against ever indexing parameters, right?

I created #119 to track this. And this is another issue that's probably worth mentioning to auditors.

apbendi commented:
@mds1 

> Issue #18 also mentions checking out solidity docgen to autogenerate docs from the natspec comments. Since we aren't doing this, we might want to create a separate issue to track documentation in general

Good call, done with #116 

> At this point 0.6.12 is a bit old. We could bump to 0.7.6 or even 0.8.2 since the latest version of @openzeppelin/contracts supports it (this means we can remove SafeMath since those checks are built in to 0.8.x, which will reduce gas cost)

At this point in the dev cycle, I'm not really comfortable jumping to a new version unless there's a clear benefit. Gas savings on SafeMath would be pretty minimal, no? My understanding is the compiler simply inlines the same kinds of checks SafeMath is already doing. So I guess maybe some savings from avoiding calls, right? If we want to gas golf I'd rather move overflow checking in manually.

> On the topic of saving gas, we could also get rid of Ownable.sol and implement it ourselves. Owable pulls in Context.sol to use _msgSender() which we don't need and makes the onlyOwner modifier + ownership transfer slightly more expensive. Ultimately _msgSender() just returns msg.sender, but still adds a little overhead. This is probably not too important of a change so I don't feel strongly here

I agree Ownable is a bit heavyweight here, but the gas savings would only come for the three admin methods, which should be called infrequently and don't impact users. Don't see a ton of benefit.

apbendi commented:
Hey @mds1, addressed your feedback and/or left responses to everything. Also, I rebased and touched up your `umbra-js` validation code to account for the updated contract parameters in the process. I'd love for you to give this whole PR one more quick pass to make sure it's looking good. Thanks!

mds1 commented:
> At this point in the dev cycle, I'm not really comfortable jumping to a new version unless there's a clear benefit. Gas savings on SafeMath would be pretty minimal, no? My understanding is the compiler simply inlines the same kinds of checks SafeMath is already doing. So I guess maybe some savings from avoiding calls, right? If we want to gas golf I'd rather move overflow checking in manually.

Regarding gas savings, here's a comment that quantifies it:  https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2465#issuecomment-756981359

Also, you could make the exact opposite argument of "we should be on the newest safe version unless there's a clear downside" 😝  I think two arguments in favor of upgrading to a more recent version are:
- Bug fixes eventually stop being ported to older versions, so we'd want to check this before deciding to stick with 0.6.x
- The optimizer has been improved a lot in more recent versions, so we may get cheaper code with a new version. Maybe we create a separate issue and use [hardhat-gas-reporter](https://github.com/cgewecke/hardhat-gas-reporter) to benchmark it + research security issues to decide about updating

> I agree Ownable is a bit heavyweight here, but the gas savings would only come for the three admin methods, which should be called infrequently and don't impact users. Don't see a ton of benefit.

This seems fine to me!

apbendi commented:
> Also, you could make the exact opposite argument of "we should be on the newest safe version unless there's a clear downside"

I think I disagree with this— a new version would carry forward any bug from the old version, but also introduce new ones and regressions. A battle tested version should be considered strictly more secure, unless there is a disclosed vulnerability in the old one, which we would know about. Of course, it's always possible a new version fixed a 0-day vulnerability in the old one by accident, but that doesn't seem worth considering.

Anyway I'm open to upgrading if it can be done without much code churn. I opened #117 to track this separately. Also worth discussing with the auditor.
